Add dataset feed to homepage

luistoptal opened the pull request:
# Pull request for issue: #199

IMPORTANT: this is a follow up to PR https://github.com/gigascience/gigadb-website/pull/2078, I suggest reviewing #2078 and this PR in combination as this one includes refactor to the changes in #2078

* Adds a dataset feed to the homepage
* Refactors the news carousel as a resuable shared partial

![image](https://github.com/user-attachments/assets/fc70a995-cd1c-4744-b625-1027b80740b4)

## How to test?

In home page http://gigadb.gigasciencejournal.com/ you should see a responsive carousel with datasets

## How have functionalities been implemented?

* the news feed from #2078 was refactored to convert the carousel into a reusable partial `protected/views/shared/_carousel_slider.php`
* In the `SiteController` a dataset search is performed and the search result passed to the view to be able to render a list of datasets as a carousel
* The number of datasets on display is limited to 12, but more can be fetched
* If more than 0 datasets are returned then render the dataset feed section `protected/views/site/datasets_carousel.php`

## Any issues with implementation?

* if the user changes the current carousel slide and then resizes the window, the carousel position is reset to slide 1. I think this is a fringe usecase so I did not handle it
* The dropdown to select type from the original example is *not implemented*, I am not sure if it is useful. If this is an important feature, it could be added
* Regarding the dataset data, I see a handful of utility methods to 'get' dataset data, e.g. `protected/components/FormattedDatasetMainSection.php`, in this PR I am directly accessing the dataset values but I do not know if I should use these utilities instead, for example, to get the dataset title i do:

```
$dataset['title']
```

But I think that it might be possible to alternatively do something like:

```
(...)
$mainSection = $assembly->getDatasetMainSection();
(...)
$mainSection->getHeadline()['title'];
```

## Any changes to automated tests?

--

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a dataset feed carousel to the homepage, showcasing published datasets with images, titles, DOIs, and publication dates.
  - Introduced a reusable, responsive carousel slider component for displaying both news and dataset items.
- **Refactor**
  - Updated the news section to use the new carousel slider for improved accessibility and layout consistency.
- **Style**
  - Added new styles for carousel slider, news section, and dataset feed section to enhance visual presentation and responsiveness.
- **Documentation**
  - Updated the changelog to reflect the addition of the dataset feed feature.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

luistoptal committed:
feat: replaces news section by responsive carousel of boxes

luistoptal committed:
refactor: adds styles to separate less file

luistoptal committed:
refactor: nests less classes

luistoptal committed:
feat: clean up news block design

luistoptal committed:
feat: enhances web accessibility

luistoptal committed:
feat: dynamically adjust visibility of carousel controls

luistoptal committed:
feat: dynamically update max width depending on number of news items to display

luistoptal committed:
feat: increase excerpt length

luistoptal committed:
refactor: move helper outside jquery wrapper since it uses pure js

luistoptal committed:
chore: up log

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal committed:
feat: abstract carousel to reusable partial

luistoptal committed:
feat: use partial in news slider

luistoptal committed:
feat: display dataseet feed in home

luistoptal committed:
feat: render datasets slides with dynamic width

luistoptal committed:
feat: simplify and improve dataset feed layout

luistoptal committed:
chore: up log

luistoptal committed:
refactor: replaces no image

luistoptal committed:
feat: improve carousel accessibility

luistoptal committed:
feat: purify dataset title

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal committed:
feat: purify news content

luistoptal committed:
feat: simplify news body line clamp using css

luistoptal committed:
chore: test news creation

luistoptal committed:
chore: remove deprecated test tag

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal committed:
chore: delete all ok-needs-secrets and ok-can-offline tags

rija commented:
Hi @luistoptal 
**chore**: a few conflicting files here, probably need rebasing before review happens

luistoptal committed:
chore: Merge branch 'develop' into new-feature/199-homepage-dataset-carousel

luistoptal commented:
Let's handle this PR after https://github.com/gigascience/gigadb-website/pull/2078 is fully reviewed and approved to avoid solving the same issue twice

luistoptal committed:
chore: update LESS imports and clean up carousel styles

- Added missing semicolon in index.less for dataset-feed-section import.
- Removed unused media queries from carousel-slider.less.
- Enhanced NewsAndFeedsService with a new method to fetch published datasets.
- Fixed potential undefined index for image tag in datasets_carousel.php.

rija reviewed (changes requested):
Hi @luistoptal,

PR #2078 has been approved and is ready for merging. 
But I won't merge it until this one has been reviewed and approved too.

**suggestion**: the aforementioned PR didn't include it, but now that the carousel is becoming a reusable component, it probably should be added to the style guide

luistoptal commented:
HI @rija I will merge the changes from #2078 into this branch and apply some necessary fixes, then set this branch as ready for review again

luistoptal committed:
chore: Merge branch 'develop' into new-feature/208-news-items-on-the-home-page

luistoptal committed:
chore: Merge branch 'new-feature/208-news-items-on-the-home-page' into new-feature/199-homepage-dataset-carousel

luistoptal committed:
chore: reapply css based line clamp and yii purification to news widget

luistoptal committed:
refactor: document and refactor carousel slider partial, split js into smaller functions

luistoptal committed:
refactor: makes throttle fnc easily reusable

luistoptal commented:
I have updated this PR as follows:

- merged latest changes from #2078 (the precursor to this PR) into this one:

* string purification has been switched to use `Yii::app()->controller->widget("CHtmlPurifier")->purify`
* news line clamp works now with CSS rather than JS, which results is a much simpler and better solution

- additional improvements:

protected/views/shared/_carousel_slider.php sees some refactor:
* added a phpdoc that describes widget usage
* move the throttle function to a separate reusable js file
* split the `arrangeSlides` function into smaller single responsibility functions to make code more readable

Style guide: now includes carousel sample

Also removed all deprecated tags from all acceptance test files (ok-can-offline, ok-needs-secrets)

luistoptal commented:
Hi @rija , now this PR includes the changes from #2078 so basically it would be possible / an option to ignore #2078 and simply review and merge this PR, if it is more convenient

luistoptal committed:
chore: add carousel to style guide

luistoptal committed:
chore: replace caoursel items by placeholders in style guide

kencho51 reviewed (commented):
Hi @luistoptal,

**praise:** The carousel of dataset image is working and the tests are passing on my dev.

**issue:** When I deployed this PR to my staging, and visited the website, I got this error:
<img width="1082" alt="image" src="https://github.com/user-attachments/assets/e62188b1-af42-4efe-8e7d-78a7d3b88d79" />

**suggestion:** The error suggests that the `/gigadb/app/client/js` is missing from the production application container, it can be fixed by adding the following line into `Production-Dockerfile`:
```
COPY ./gigadb/app/client/js/ /gigadb/app/client/js/
```

**issue:** The carousel is not showing the complete image in my staging as below:
![image](https://github.com/user-attachments/assets/f0ed1e3f-7433-49bf-a99d-05f2295946ca)
**question:** Is is possible to make the carousel to show a complete image?

**chore:** Fix the merge conflicts.

only1chunts commented:
> **issue:** The carousel is not showing the complete image in my staging as below: ![image](https://private-user-
> **question:** Is is possible to make the carousel to show a complete image?
 

Hi, Just to say, yes this definately needs fixing before release. If it helps we do try to follow some rules on image sizing to ensure the thumbnail images are all pretty much the same size. i.e. they should all be square images (not rectangles) and they should be approx 400x400 pixels.
If we ever get around to completing ticket #1295 (or its replacement as that one was closed without the criteria being fulfilled?!) then the older thumbnail images may be more variable in size, but we can look to adjust those if required.

luistoptal committed:
fix: copy app client js to production

luistoptal committed:
fix: dataset feed shows full non trimmed images

luistoptal committed:
chore: Merge branch 'develop' into new-feature/199-homepage-dataset-carousel

luistoptal commented:
hi @kencho51 

I updated as per your suggestions. To keep in mind, if an image is not square, it will not fill the entire space available, e.g.

![image](https://github.com/user-attachments/assets/a5ac5da6-94de-478e-8530-98ced11116d3)

I cannot yet deploy to my own staging though, to be able to cross check. Can you tell me if it works now? By the way, the repo contains a handful of `Production-Dockerfile`, I assume the correct one is `ops/packaging/Production-Dockerfile`

kencho51 reviewed (approved):
Hi @luistoptal,

**praise:** The carousel feed is working my staging now, and the non-squared  image cannot fully fill seems not a big problem to me as below:
![image](https://github.com/user-attachments/assets/5319a19c-6db5-4c46-9029-96156dfe8b26)

**notes:** This PR is combined with PR #2078, which I have already approved.

**chore:** Fix the changelog conflicts.

**praise:** All tests are passing.

LGTM, happy to approve (assuming the chore has been fixed).

rija reviewed (changes requested):
Hi @luistoptal 

**praise**: functionality works and looks gorgeous
**praise**: style guide has been updated
**issue**: In the PHP code for the new service, methods are missing type hints
**suggestion**: Add type hints. You can follow the example from another service: `protected/components/FiledropService.php`, line 168 for the `emailInstructions` method
**praise**: all tests are passing

luistoptal committed:
refactor: update getFeedDatasets method signature to include type hints for better clarity

luistoptal committed:
chore: Merge branch 'develop' into new-feature/199-homepage-dataset-carousel

luistoptal commented:
Hi @rija 
I have applied your suggestions and solved conflicts

luistoptal committed:
chore: Merge branch 'develop' into new-feature/199-homepage-dataset-carousel

luistoptal commented:
@rija merge conflicts are solved

rija reviewed (approved):
Hi @luistoptal 

happy to approve
